File: modules/widgets/lib/rin/widgets/containers/WCRoot.cpp
#include "rin/widgets/containers/WCRoot.hpp"

#include "rin/widgets/ContainerWidgetSlot.hpp"

Vec2<float> WCRoot::ComputeContentSize()
{
    return Vec2{0.0f};
}

void WCRoot::ArrangeSlots(const Vec2<float>& drawSize)
{
    for(const auto &slot : GetSlots())
    {
        const auto widget = slot->GetWidget();
        widget->SetOffset(Vec2{0.0f});
        widget->SetSize(drawSize);
    }
}

File: rin.Widgets/Containers/WCRoot.cs
namespace rin.Widgets.Containers;

public class WCRoot : Container
{
    protected override Size2d ComputeDesiredContentSize()
    {
        return new Size2d();
    }

    protected override void ArrangeSlots(Size2d drawSize)
    {
        foreach (var slot in GetSlots())
        {
            OnSlotUpdated(slot);
        }
    }

    public override void OnSlotUpdated(Slot slot)
    {
        base.OnSlotUpdated(slot);
        var widget = slot.GetWidget();
        widget.SetOffset(0.0f);
        widget.SetSize(GetSize());
    }

    public override void SetSize(Size2d size)
    {
        base.SetSize(size);
        ArrangeSlots(GetContentSize());
    }
}